### PR TITLE
[Feature / Option] Vorticity Confinement (VC) technique to reduce numerical diffusion

### PR DESCRIPTION
**Pull request by josy-nal**
## Proposed Changes
*Give a brief overview of your contribution here in a few sentences.*
 
Implemented the option for vorticity confinement technique as a feature branch on the latest su2code/SU2:develop branch.

Descretized schemes offer numerical diffusion which results in lower resolution and accuracy in solution. This issue is especially observed in vortex dominated flows. A fine mesh will be necessary to get a solution with good accuracy. This increases the computation time and cost. A better technique is required which provides reasonably good accuracy without increasing the computation time and costs much.

Vorticity confinement technique is a proven procedure that helps in reducing numerical diffusion offered by the scheme. Results from the literature shows that using the vorticity confinement technique, a better resolution is achieved in the problems like flow behind bluff bodies, rotors, wings etc.

References

J. Steinhoff and D. Underhill, ``Modification of the euler equations for vorticity confinement: Application to the computation of interacting vortex rings,'' Physics of Fluids, vol.6, no.8, pp. 2738--2744, 1994.  https://doi.org/10.1063/1.868164

D. A. Caughey and M. M. Hafez, Frontiers of Computational Fluid Dynamics 2006.  WORLD SCIENTIFIC, 2005.  https://www.worldscientific.com/doi/abs/10.1142/5963

R. Loehner and C. Yang, Vorticity confinement on unstructured grids, 40th AIAA Aerospace Sciences Meeting &amp; Exhibit, 2002. [Online]. Available: https://arc.  aiaa.org/doi/abs/10.2514/6.2002-137

A. Povitsky and K. C. Pierson, ``Vorticity confinement applied to accurate prediction of convection of wing tip vortices and induced drag,'' International Journal of Computational Fluid Dynamics, vol. 35, no. 3, pp. 143--156, 2021.  https://doi.org/10.1080/10618562.2020.1856822

## Related Work
*Resolve any issues (bug fix or feature request), note any related PRs, or mention interactions with the work of others, if any.*

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [X] I am submitting my contribution to the develop branch.
- [X] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [ ] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [X] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

### CODE REVIEW

**Review by chanduyal — COMMENTED** [review]
As a contributor, I'm fine with this pull request.

File: SU2_CFD/src/solvers/CEulerSolver.cpp [hunk 1]
```diff
@@ -2215,6 +2216,74 @@ void CEulerSolver::Source_Residual(CGeometry *geometry, CSolver **solver_contain
...
+      // set vorticity magnitude as auxilliary variable
+      const su2double VorticityMag = max(GeometryToolbox::Norm(3, nodes->GetVorticity(iPoint)), 1e-12);
+      nodes->SetAuxVar(iPoint, 0, VorticityMag);
+    }
+    SU2_MPI::Allreduce(&VolumeSum_local, &VolumeSum_global, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+    const unsigned long nPointLocal = geometry->GetnPoint();
+    unsigned long nPointGlobal = 0;
+    SU2_MPI::Allreduce(&nPointLocal, &nPointGlobal, 1, MPI_UNSIGNED_LONG, MPI_SUM, MPI_COMM_WORLD);
+    const su2double AvgVolume = VolumeSum_global / nPointGlobal;
+    second_numerics->SetAvgVolume_VC(AvgVolume);
```
Comment (pcarruscag) [inline SU2_CFD/src/solvers/CEulerSolver.cpp:2265]: `GetGlobal_nPointDomain` is not being computed for coarse multigrid levels, please modify CMultigridGeometry.cpp around line 510 to store the local variable `Global_nPointCoarse`
```suggestion
    SU2_OMP_FOR_STAT(omp_chunk_size)
    for (iPoint = 0; iPoint < nPoint; iPoint++) {
      /*--- Set vorticity magnitude as the auxiliary variable. ---*/
      const su2double VorticityMag = max(GeometryToolbox::Norm(3, nodes->GetVorticity(iPoint)), 1e-12);
      nodes->SetAuxVar(iPoint, 0, VorticityMag);
    }
    END_SU2_OMP_FOR

    const su2double AvgVolume = config->GetDomainVolume() / geometry->GetGlobal_nPointDomain();
    second_numerics->SetAvgVolume_VC(AvgVolume);
```

File: SU2_CFD/src/solvers/CEulerSolver.cpp [hunk 2]
```diff
@@ -2215,6 +2216,74 @@ void CEulerSolver::Source_Residual(CGeometry *geometry, CSolver **solver_contain
...
+
+    SetAuxVar_Gradient_GG(geometry, config);
+
+    SU2_OMP_FOR_DYN(omp_chunk_size)
+    for (iPoint = 0; iPoint < nPointDomain; iPoint++) {
+      second_numerics->SetConservative(nodes->GetSolution(iPoint), nodes->GetSolution(iPoint));
+      second_numerics->SetPrimitive(nodes->GetPrimitive(iPoint), nodes->GetPrimitive(iPoint));
+      second_numerics->SetVorticity(nodes->GetVorticity(iPoint), nodes->GetVorticity(iPoint));
+      second_numerics->SetAuxVarGrad(nodes->GetAuxVarGradient(iPoint), nullptr);
+      second_numerics->SetWallDistance_Numerics_VC(geometry->nodes->GetWall_Distance(iPoint));
```
Comment (pcarruscag) [inline SU2_CFD/src/solvers/CEulerSolver.cpp:2275]: After the other suggestions it should look like this.
```suggestion
      second_numerics->SetPrimitive(nodes->GetPrimitive(iPoint), nullptr);
      second_numerics->SetVorticity(nodes->GetVorticity(iPoint), nullptr);
      second_numerics->SetAuxVarGrad(nodes->GetAuxVarGradient(iPoint), nullptr);
      second_numerics->SetDistance(geometry->nodes->GetWall_Distance(iPoint), 0);
```

File: TestCases/rans/oneram6/turb_ONERAM6.cfg [hunk 3]
```diff
@@ -25,6 +25,13 @@ FREESTREAM_TEMPERATURE= 288.15
 REYNOLDS_NUMBER= 11.72E6
 REYNOLDS_LENGTH= 0.64607
 
+% --------------------------- VORTICITY_CONFINEMENT ---------------------------%
+%
+% Enable vorticity confinement (YES/NO)
+VORTICITY_CONFINEMENT = NO
+% Set confinement parameter (0.00 by default)
+CONFINEMENT_PARAM = 0.00
+
```
Comment (pcarruscag) [inline TestCases/rans/oneram6/turb_ONERAM6.cfg:34]: Can you exercise this feature in one of the existing regression tests?

**Review by pcarruscag — COMMENTED** [review]
Welcome to SU2 and thank you for this contribution, it looks good, minor cleanups and simplifications are suggested below.
Do you have some case where this was particularly useful that you could add to our V&V collection (https://su2code.github.io/vandv/home/)?

Is this technique also applicable to incompressible flows? If so please consider adding it to CIncEulerSolver too.

In CConfig::SetPosprocessing you should check for features that exclude the use of this source term and throw an error. Currently is it not compatible with the:
- Incompressible
- NEMO
- Axisymmetry

**Comment by pcarruscag** [comment]
Ah yes, please add yourselves to AUTHORS.md

File: SU2_CFD/src/numerics/flow/flow_sources.cpp [hunk 4]
```diff
@@ -616,6 +616,130 @@ CNumerics::ResidualType<> CSourceIncRotatingFrame_Flow::ComputeResidual(const CC
...
+  const su2double nXomega_U = (nXomega_x *u) + (nXomega_y *v) + (nXomega_z *w);
+
+  // vorticity confinement parameter
+  const su2double vc_config = config->GetConfinement_Param();
+  su2double vc = vc_config;
+
+  const su2double one_third = 0.33333333; // 1./3.;
+  vc *= (1 + log10(pow((1+(Volume/AvgVolume)), one_third)));
+
+  // correction to vc near viscous wall
```
Comment (pcarruscag) [inline SU2_CFD/src/numerics/flow/flow_sources.cpp:675]: Our comment style (for better or worse) is:
```suggestion
  /*--- correction to vc near viscous wall ---*/
```

File: SU2_CFD/src/numerics/flow/flow_sources.cpp [hunk 5]
```diff
@@ -616,6 +616,106 @@ CNumerics::ResidualType<> CSourceIncRotatingFrame_Flow::ComputeResidual(const CC
...
+CSourceVorticityConfinement::CSourceVorticityConfinement(unsigned short val_nDim, unsigned short val_nVar,
+                                                         const CConfig* config)
+    : CSourceBase_Flow(val_nDim, val_nVar, config) {
+  Gamma = config->GetGamma();
+  Gamma_Minus_One = Gamma - 1.0;
+}
+
+CNumerics::ResidualType<> CSourceVorticityConfinement::ComputeResidual(const CConfig* config) {
+  /*--- density, \rho ---*/
+  const su2double rho = U_i[0];
```
Comment (pcarruscag) [inline SU2_CFD/src/numerics/flow/flow_sources.cpp:629]:
```suggestion
CSourceVorticityConfinement::CSourceVorticityConfinement(unsigned short val_nDim, unsigned short val_nVar,
                                                         const CConfig* config)
    : CSourceBase_Flow(val_nDim, val_nVar, config) {
  Gamma = config->GetGamma();
  Gamma_Minus_One = Gamma - 1.0;
}

CNumerics::ResidualType<> CSourceVorticityConfinement::ComputeResidual(const CConfig* config) {
  /*--- density, \rho ---*/
  const su2double rho = V_i[nDim + 2];
```

File: SU2_CFD/src/solvers/CEulerSolver.cpp [hunk 6]
```diff
@@ -2215,6 +2216,43 @@ void CEulerSolver::Source_Residual(CGeometry *geometry, CSolver **solver_contain
...
+      nodes->SetAuxVar(iPoint, 0, VorticityMag);
+    }
+    END_SU2_OMP_FOR
+
+    /*--- calculate the gradient of the vorticity magnitude (AuxVarGradient) ---*/
+    SetAuxVar_Gradient_GG(geometry, config);
+
+    SU2_OMP_FOR_DYN(omp_chunk_size)
+    for (iPoint = 0; iPoint < nPointDomain; iPoint++) {
+      second_numerics->SetConservative(nodes->GetSolution(iPoint), nullptr);
```
Comment (pcarruscag) [inline SU2_CFD/src/solvers/CEulerSolver.cpp:2240]:
```suggestion
```

File: SU2_CFD/src/solvers/CEulerSolver.cpp (hunk 1 above)
Comment (pcarruscag) [reply SU2_CFD/src/solvers/CEulerSolver.cpp]: You need to make the change to CMultigridGeometry.cpp that I mention above

**Comment by pcarruscag** [comment]
Hi, thank you for making the changes, please have another look at the questions in the main comment from my review + my comment from today + my older comment regarding AUTHORS.md

File: AUTHORS.md [hunk 7]
```diff
@@ -152,5 +152,8 @@ sravya91
 srcopela
 tobadavid
 vfrancesmolla
+Y Chandukrishna
+Josy P Pullockara
+T N Venkatesh
```
Comment (pcarruscag) [inline AUTHORS.md:157]: Alphabetical order please, which in this list happens to put upper case before lower case.

File: TestCases/rans/oneram6/turb_ONERAM6.cfg [hunk 8]
```diff
@@ -25,6 +25,13 @@ FREESTREAM_TEMPERATURE= 288.15
 REYNOLDS_NUMBER= 11.72E6
 REYNOLDS_LENGTH= 0.64607
 
+% --------------------------- VORTICITY_CONFINEMENT ---------------------------%
+%
+% Enable vorticity confinement (YES/NO)
+VORTICITY_CONFINEMENT = NO
+% Set confinement parameter (0.00 by default)
+CONFINEMENT_PARAM = 0.00
```
Comment (pcarruscag) [inline TestCases/rans/oneram6/turb_ONERAM6.cfg:33]: Thank you for the changes, you just need to add a regression test to parallel_regression.py that uses the new feature.

**Review by pcarruscag — APPROVED** [review]

**Comment by tnv-nal** [comment]
Thank you

**Pull request merged**